Sources/GraphQLWebSocket/Client.swift

```diff
@@ -0,0 +1,583 @@
+                return "acknowledged"
+            case .notconnected:
+                return "notconnected"
+            case .connecting:
+                return "connecting"
+            case .connected:
+                return "connected"
+            case .reconnecting(retry: let retry):
+                return "reconnecting (\(retry))"
+            case .disposed:
+                return "disposed"
+            }
+        }
+    }
+    
+    /// Queue of messages that should be sent to the server.
+    ///
+    /// - NOTE: The 0-th message should be sent first.
+    private var queue = [ClientMessage]()
+    
+    public enum Event {
+        /// Client started connecting.
+        case connecting
+        
+        /// WebSocket has opened.
+        case opened(socket: WebSocket)
+        
+        /// Open WebSocket connection has been acknowledged
+        case acknowledged(payload: [String: AnyCodable]?)
+        
+        /// Ping message has been received or sent.
+        ///
+        /// - parameter received: Tells whether the ping was received from the server. If `false` the client sent the ping.
+        case ping(received: Bool, payload: [String: AnyCodable]?)
+        
+        /// Pong message has been received or sent.
+        ///
+        /// - parameter received: Tells whether the pong was received from the server. If `false` the client sent the event.
+        case pong(received: Bool, payload: [String: AnyCodable]?)
+        
+        /// A message from the server has been received.
+        case message(ServerMessage)
+        
+        /// WebSocket connection has closed.
+        case closed
+        
+        /// WebSocket connection had an error or client had an internal error.
+        case error(Error)
+    }
+    
+    /// The central subject that publishes all events to the pipelines.
+    private let emitter = PassthroughSubject<Event, Never>()
+    
+    /// A timer reference responsible for checking that the server has ACK the connection in timely manner.
+    /// https://stackoverflow.com/a/26808801/2946444
+    weak private var ackTimer: Timer?
+    
+    /// A timer reference responsible for keeping the connection alive by sending pings.
+    /// https://stackoverflow.com/a/26808801/2946444
+    weak private var pingTimer: Timer?
+    
+    /// A timer reference responsible for retrying to reconnect after a given period of time elapses.
+    /// https://stackoverflow.com/a/26808801/2946444
+    weak private var reconnectTimer: Timer?
+    
+    /// Timer that starts the disconnect procedure when there are no more listeners.
+    weak private var lazyCloseTimer: Timer?
+    
+    /// Timer that fires off when the server hasn't replyed with PONG for too long.
+    weak private var connectionDroppedTimer: Timer?
+    
+    /// Holds references to pipelines created by subscriptions. Each pipeline is identified by the
+    /// query id of the subscription that created a pipeline.
```

Comment:
```suggestion
    /// Holds references to pipelines created by subscriptions. Each pipeline is identified by the
    /// random id of the subscription that created a pipeline.
```
The id is a UUID() and not something that comes from the query

Reply:
I see your point, but I think it's important to point out that this same `id` that's being used to identify the pipeline also appears in the `id` parameter of the server message.